language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -55,54 +98,134 @@ type VarEnv = UniqueMap TermUnique Closure
 data CekEnv = CekEnv
     { _cekEnvMeans  :: DynamicBuiltinNameMeanings
     , _cekEnvVarEnv :: VarEnv
+    , _cekEnvProtocolParameters :: ()
     }
 
+data ExBudget = ExBudget { _exBudgetCPU :: ExCPU, _exBudgetMemory :: ExMemory }
+  deriving (Eq, Show)
+
+data CekState w s = CekState
+    { _cekStateWriter :: w -- ^ for counting what cost how much
+    , _cekStateState :: s  -- ^ for making sure we don't spend too much
+    }
+
+data ExTally = ExTally
+    { _exTallyCPU :: MonoidalHashMap (Plain Term) [ExCPU]
+    , _exTallyMemory :: MonoidalHashMap (Plain Term) [ExMemory]
+    }
+    deriving stock (Eq, Generic, Show)
+    deriving (Semigroup, Monoid) via (GenericSemigroupMonoid ExTally)
+
+$(join <$> traverse makeLenses [''CekEnv, ''CekState, ''ExBudget, ''ExTally])
```

Comment:
I'd prefer a bunch of lines with `makeLenses` for each of the data types, but no strong feelings. Note also that we have this definition:

```haskell
forBind :: (Monad m, Traversable m, Applicative f) => m a -> (a -> f (m b)) -> f (m b)
forBind a f = join <$> traverse f a
```

in `Control.Monad.Trans.Inner`. (better name suggestions are welcome)

Reply:
Faster this way, because the compiler doesn't need to generate a new TH context each time.

Reply:
Good point. Do you happen to have a reference? Just curious how different the numbers are.

Reply:
Not too relevant. I've once used it for about 20 data structures + lenses, and there it was slightly noticeable (think 0.5s faster).